File: ocaml/src/p0062.ml
(* Project Euler: Problem 62 *)

open Core

let compute qty =
  let make_key n = Euler.Util.digits n |> List.sort ~compare |> Euler.Util.undigits in
  let tbl = Hashtbl.create (module Int) in
  let rec loop n =
    let cube = Int.pow n 3 in
    let cube_lst =
      Hashtbl.update_and_return tbl (make_key cube) ~f:(fun v ->
        match v with
        | None -> [ cube ]
        | Some lst -> cube :: lst)
    in
    if List.length cube_lst = qty then List.last_exn cube_lst else loop (succ n)
  in
  loop 1
;;

let solve () = compute 5 |> Int.to_string

(* Test *)

let%test_unit "3" = [%test_eq: int] (compute 3) 41063625
let%test_unit "5" = [%test_eq: int] (compute 5) 127035954683

File: rust/src/bin/p0062.rs
// Project Euler: Problem 62

euler::run_solver!(62);

fn solve() -> String {
    compute(5).to_string()
}

fn compute(n_perms: usize) -> u64 {
    use euler::math;
    use std::collections::HashMap;

    fn make_key(n: u64) -> u64 {
        let mut tmp = math::digits(n);
        tmp.sort_unstable();
        math::undigits(&tmp)
    }

    debug_assert!(n_perms > 1);

    let mut tbl: HashMap<u64, Vec<u64>> = HashMap::new();
    for n in 1_u64.. {
        let cube = n.pow(3);
        let key = make_key(cube);
        tbl.entry(key).or_default().push(cube);
        let tmp = tbl.get(&key).unwrap();
        if tmp.len() == n_perms {
            return tmp[0];
        }
    }

    unreachable!();
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0062_3() {
        assert_eq!(compute(3), 41063625);
    }

    #[test]
    fn p0062_5() {
        assert_eq!(compute(5), 127035954683);
    }
}
